### ISSUE TITLE
Promise needs to be resolved synchronously for property to fail

### ISSUE THREAD

**Issue by Ezku** [jsverify/jsverify#87]
Hi,

The documentation mentions support for Promises, but it does seem to ship with corner cases that made things a bit difficult.

This property fails as it should:

```coffeescript
jsc.property 'foo', ->
   Promise.resolve(false).delay(10)
```

This also fails as it should:

```coffeescript
jsc.property 'foo', jsc.constant('bar'), ->
   Promise.resolve(false)
```

However, this last variation passes, which shouldn't happen:

```coffeescript
jsc.property 'foo', jsc.constant('bar'), ->
  Promise.resolve(false).delay(10)
```

Also, only the first iteration seems to be ran.

I circumvented this by writing my own combinator:

```coffeescript
property = (description, args...) ->
  it description, ->
    jsc.check jsc.forall(args...)
```

However, I get `OK, passed 100 tests` in the middle of my test output which I'd like to avoid. You'll also likely want to look into _why_ `jsc.property` acts the way it does.

Awesome work on jsverify, I'm enjoying my trip so far.

Thanks!

**Comment by Ezku** [comment]
PS. I'm running 0.6.0-alpha.3. Sorry, didn't have time to check whether this affects other versions.

**Comment by phadej** [comment]
Seems I have used `jsc.assert(...).done()` myself. Have to check what I broke in between.

Luckily it's only alpha!

**Comment by phadej** [comment]
Oh wait, I doesn't `done`. Could you write how you use `chai`. IIRC it throws exceptions, and those doesn't work?

**Comment by Ezku** [comment]
Here's a combinator that seems to be failing and succeeding asynchronously as should.

```coffeescript
property = (description, args...) ->
  it description, (done) ->
    jsc.check(jsc.forall(args...)).then (holds) ->
      if holds is true
        done()
      else
        done new Error "property does not hold, counterexample: #{holds.counterexamplestr}"
```

A problem I didn't see immediately was that `holds` is either boolean true _or_ an object describing the counterexample.

**Comment by phadej** [comment]
Sorry I don't understand?  `jsc.property` eats exception?

Is the minimum not failing example:
```js
jsc.property("fails", "nat", function () {
  return delay(promise).then(function () {
    throw new Error("fail always");
  });
});
```

**Comment by phadej** [comment]
Could you try if https://github.com/jsverify/jsverify/pull/88 issue-87 branch fixes your use case.

**Comment by phadej** [comment]
Hopefully fixed in `0.6.0-alpha.4`. Could you verify?

**Comment by Ezku** [comment]
Updating to `0.6.0-alpha.4` fixes my use case. Solid work, thank you!
